Emit analytics events when users vote on forum posts.

smarnach opened the pull request:
**Description**
The forum client already emits analytics events when forum posts are created.  This PR adds analytics events when a user votes on a thread or response.

**Open issues**
- [x] Add tests

**Sandbox:** [LMS](http://pr9616.sandbox.opencraft.com/), [Studio](http://studio.pr9616.sandbox.opencraft.com/)

edx-webhook commented:
Thanks for the pull request, @smarnach! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=9616&repo=edx%2Fedx-platform

bradenmacdonald commented:
Seems to work well! Will approve once you add some tests and get a passing build. We should also ask about the status of downvotes and find out if they will be removed or implemented more fully at some point.

smarnach commented:
@bradenmacdonald I added tests for the events and a few docstrings.  The bokchoy failure I got on the last build was probably a flaky test, and the quality failure probably caused by the missing docstring, so I hope for a green build now.

We can ask about downvotes during upstream review.

bradenmacdonald commented:
@smarnach There are still a few minor quality issues:

```
21:57:19 lms/djangoapps/django_comment_client/base/views.py (95.3%):
21:57:19     148: C0103: (invalid-name), permitted: Invalid argument name "fn"
21:57:19     148: C0111: (missing-docstring), permitted: Missing function docstring
21:57:19     150: C0111: (missing-docstring), permitted.wrapper: Missing function docstring
21:57:19     151: C0111: (missing-docstring), permitted.wrapper.fetch_content: Missing function docstring
21:57:19     169: C0111: (missing-docstring), ajax_content_response: Missing function docstring
21:57:19     467: C0111: (missing-docstring), _vote_or_unvote: Missing function docstring
```

smarnach commented:
@bradenmacdonald Do you think this is ready for upstream review now?  I just rebased on top of current master.

bradenmacdonald commented:
@smarnach Yep, should be good. :+1: Go ahead and open an OSPR.

edx-webhook commented:
Thanks for the pull request, @smarnach! I've created [OSPR-813](https://openedx.atlassian.net/browse/OSPR-813) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

sarina commented:
@smarnach have you talked with our analytics team about these new events & their format?

smarnach commented:
@sarina No, we haven't yet.

sarina commented:
@smarnach in the future, I think you guys should consider engaging with the analytics team in the design process. It will make things much simpler. This will need review by both Analytics and T&L. I think it makes sense for analytics to be first, to review the format, then T&L can review the changes to the forums code.

smarnach commented:
@sarina Thanks for the recommendation.  We'll keep this in mind when adding analytics events again.

bradenmacdonald commented:
@sarina Just to clarify, we did mention most of our recent work in advance [on the mailing list](https://groups.google.com/d/msg/openedx-analytics/DAJ2d7Jq2s4/Lq8uKY0LC4kJ) and in quite a few emails between myself and Victor, but due to the volume of work and the short time schedule we unfortunately didn't have the time to do a deep design dive and back-and-forth design review on every task before implementation. I agree that would be preferable, obviously.

sarina commented:
Braden, I totally get that, but they won't be merged upstream unitl that
conversation is had, I don't think. What I'm suggesting is to have that
conversation prior to the upstream pull request.

On Tue, Sep 15, 2015 at 1:39 PM, Braden MacDonald <notifications@github.com>
wrote:

> @sarina <https://github.com/sarina> Just to clarify, we did mention most
> of our recent work in advance on the mailing list
> <https://groups.google.com/d/msg/openedx-analytics/DAJ2d7Jq2s4/Lq8uKY0LC4kJ>
> and in quite a few emails between myself and Victor, but due to the volume
> of work and the short time schedule we unfortunately didn't have the time
> to do a deep design dive and back-and-forth design review on every task
> before implementation. I agree that would be preferable, obviously.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/9616#issuecomment-140477218>.
>

bradenmacdonald commented:
@sarina Ok, makes sense.

shnayder commented:
Hi @smarnach. Thanks for the PR! We definitely want to track votes. To make it easier to review, can you paste in an example of each new event (not sure if there's more than one) in the PR description?

smarnach commented:
@shnayder Sorry for the delay in replying.  Example events are available at https://gist.github.com/smarnach/ad8ccab974a585fe9253.  There are two events of type `edx.forum.thread.voted` and two events of type `edx.forum.response.voted`, one vote and one unvote each.  I also rebased on top of current master.

lamagnifica commented:
tagging myself for doc: @lamagnifica

dianakhuang commented on lms/djangoapps/django_comment_client/base/views.py:
```diff
@@ -456,20 +480,36 @@ def delete_comment(request, course_id, comment_id):
+    """
+    Vote or unvote for a thread or a response.
+    """
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
```
It's recommended that you use `CourseKey.from_string` instead of `SlashSeparatedCourseKey.from_deprecated_string`.

smarnach replied:
@dianakhuang I didn't add this line – I just moved it here.  There are about 20 more places in the same file that use `SlashSeparatedCourseKey.from_deprecated_string()`, and not a single one using `CourseKey.from_string()`, so changing this one occurrence would feel weird.  Changing all of them seems outside the scope of this PR and would be better done in a separate PR.

sarina replied:
Since this has been uncovered, it should be fixed. This should apply to all techdebt - we enforce it with pep8/pylint, but we should all improve the codebase whenever we touch older areas. If you feel it's more appropriate to do so in a separate PR, then do that.

smarnach replied:
@sarina I've posted #10660 to clean this up.  It can only be merged after this PR, since I had to base it on the topic branch for this PR.  If there are further changes necessary to this PR, they need to be forward-ported.

dianakhuang commented on lms/djangoapps/django_comment_client/base/views.py:
```diff
@@ -139,29 +111,84 @@ def get_thread_created_event_data(thread, followed):
+    if isinstance(obj, cc.Thread):
+        obj_type = 'thread'
+    else:
+        obj_type = 'response'
```
Is there a reason why we only have `'response'` and not `'comment'` here as well?

smarnach replied:
Yes, the reason is that you can't vote on comments.  You can only vote on threads and responses (and this is reflected everywhere in the code, not just here, so it doesn't make sense to change it here).

dianakhuang commented on lms/djangoapps/django_comment_client/base/views.py:
```diff
@@ -456,20 +480,36 @@ def delete_comment(request, course_id, comment_id):
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    course = get_course_with_access(request.user, 'load', course_key)
+    user = cc.User.from_django_user(request.user)
+    if value == 'undo':
```
I understand why you're doing things this way, but I'd prefer it if we didn't put `undo` into the value and instead add a separate `undo_vote` parameter to `_vote_or_unvote`, especially since `undo_vote` is a separate parameter to `track_voted_event`. I think it will just make this logic a lot clearer all around.

smarnach replied:
Thanks, that's a good suggestion.  The reason for the current version was that there wouldn't be a useful value for `value` when undoing votes.

dianakhuang commented on lms/djangoapps/django_comment_client/base/tests.py:
```diff
@@ -1642,6 +1642,40 @@ def test_team_events(self, view_name, event_name, view_data, view_kwargs, mock_r
+        ('undo_vote_for_comment', 'comment_id', 'response'),
+    )
+    @ddt.unpack
+    @patch('eventtracking.tracker.emit')
```
I understand that you're just updating an older test file to test your particular function, but just for future reference, there is a [useful event testing mixin](https://github.com/edx/edx-platform/blob/master/common/djangoapps/util/testing.py#L62) for any future events testing you might end up doing.

smarnach replied:
@dianakhuang Thanks for the pointer, good to know.  I briefly looked into porting the current code over, but that would be a bit cumbersome, since I would have to port over all existing test cases (and the current assertion lists are far from being complete, so they would need to be updated as well).

dianakhuang commented:
That's my first pass. Overall, it's looking great! Thanks for cleaning up the thread/comment creation events.

brianhw commented on lms/djangoapps/django_comment_client/base/views.py:
```diff
@@ -139,29 +111,84 @@ def get_thread_created_event_data(thread, followed):
+            else:
+                content = None
+            return content
+        course_key = SlashSeparatedCourseKey.from_deprecated_string(kwargs['course_id'])
```
Is this always in a deprecated form?  Thought recommendation was to just use CourseKey.from_string(), which should support both.

brianhw replied:
Yes, I thought I saw this before -- Diana already made this point below.

smarnach replied:
@brianhw There are roughly 20 occurrences of this in this file, and none of them was added by me.  I just moved the lines around.  I don't think it makes sense to clean this up in this PR.

smarnach replied:
Sarina requested to clean this up now, so I posted #10660, since I feel it doesn't belong in this PR at all.  (It would make the diff even less readable.)

brianhw commented:
From an analytics perspective, the code looks reasonable as does the output (in the gist).  As for testing, this seems to be one of three files that patch the tracker directly instead of using the mixin Diana mentioned.  Again, not a blocker.

explorerleslie commented:
@smarnach have you had a chance to review the feedback on this PR? Can you give us a timeline for when this will be ready for a re-review with feedback addressed? cc @sarina

smarnach commented:
@dianakhuang @brianhw Thanks for the review.  I addressed your comments and rebased on top of current master.  Could you please take another look?

smarnach commented:
jenkins run bokchoy

dianakhuang commented:
Thanks for the update! :+1: from me. I'll take a look at your cleanup PR now.

sarina commented:
@dianakhuang can @smarnach go ahead and merge this, once he squashes the commits up (you should squash out that fixup)

dianakhuang commented:
@sarina Yup, I think it's ready to merge.

smarnach commented:
@dianakhuang @sarina Thank you!  I've rebased and squashed.  Merging now even though the build hasn't finished again, since it was green before rebasing and I don't to wait until after the Django-1.8 merge.  (I've run the relevant Python tests locally after rebasing.)

stroilova commented:
Hi @smarnach  Thank you again for contributing these events. They'll be very valuable for the research community.

Quick question about format -- we were looking over the output sample events provided in the gist while writing researcher docs, and noticed that the commentable_id field seems to be set to 'course' instead of a hash of the component or an i4x identifier for the top level forum category
(See docs here: https://edx.readthedocs.org/en/latest/internal_data_formats/discussion_data.html?highlight=commentable_id#commentable-id )

For example, edx.forum.response.created events have a field like:
commentable_id: 792b252345620e6ee8180bde50f76fea773a9d06 

In the gist, I was seeing  "commentable_id" : "course" instead. Is this intended?

Also, you mentioned in the Docs PR that the main code has been updated, and events now include a team_id in common fields. Have you quickly verified your events on localhost / sandbox to make sure that other things have not changed?

I can also help with verifying events on Stage and Prod after they are released. 

Many thanks again.

FYI: @dianakhuang  @sarina

smarnach commented:
@stroilova The `commentable_id` can either be the id of a particular component the discussion belongs to for inline discussions, or it can have the value `"course"` for the global discussion forum in the "Discussion" tab.  I didn't change the way `commentable_id` is set – it's set in the same way for all forum events.

I don't have the time to test this again on the devstack at the moment, so I took another brief look at the code, and it looks like nothing else has changed.  The `team_id` is only set when this actually is a team discussion.  There are also the fields `category_name` and `category_id`, which will be shown for inline discussions only.  They also apply to all forum events, and I don't know their exact meaning, since I didn't introduce them.

stroilova commented:
Thank you very much @smarnach  -- this looks great.
